src/formatters/utils/sortResults.ts

```diff
@@ -1,13 +1,79 @@
 import { IRuleResult } from '../../types';
 
-export const sortResults = (results: IRuleResult[]) => {
-  return [...results].sort((resultA, resultB) => {
-    const diff = resultA.range.start.line - resultB.range.start.line;
+const compareCode = (one: string | number | undefined, another: string | number | undefined): number => {
+  if (one === void 0 && another === void 0) {
+    return 0;
+  }
+
+  if (one === void 0) {
+    return -1;
+  }
+
+  if (another === void 0) {
+    return 1;
+  }
+
+  return one.toString().localeCompare(another.toString(), undefined, { numeric: true });
+};
+
+const compareSource = (one: string | undefined, another: string | undefined): number => {
+  if (one === void 0 && another === void 0) {
+    return 0;
+  }
+
+  if (one === void 0) {
+    return -1;
+  }
 
-    if (diff === 0) {
-      return resultA.range.start.character - resultB.range.start.character;
-    }
+  if (another === void 0) {
+    return 1;
+  }
+
+  return one.localeCompare(another);
+};
+
+const normalize = (value: number): -1 | 0 | 1 => {
+  if (value < 0) {
+    return -1;
+  }
+
+  if (value > 0) {
+    return 1;
+  }
+
+  return 0;
+};
 
-    return diff;
-  });
+export const resultsComparer = (resultA: IRuleResult, resultB: IRuleResult): -1 | 0 | 1 => {
+  const diffSource = compareSource(resultA.source, resultB.source);
```

Comment:
Also, how do you feel about having a sort of loop here?
The performance difference will be negliable, but the code will look nicer, IMHO.
```ts
return normalize([
  compareSource(resultA.source, resultB.source),
  resultA.range.start.line - resultB.range.start.line,
  resultA.range.start.character - resultB.range.start.character,
  compareCode(resultA.code, resultB.code),
  resultA.path.join().localeCompare(resultB.path.join()),
].find(diff => diff !== 0) ?? 0); // ?? 0 just in case all previous calls returned 0
```
Didn't test it at all, but it should work.

Reply:
@P0lip That should work. But I feel a bit wary about this kind of code. It's usually very painful to debug/troubleshoot when it doesn't behave as expected.

For some reasons that reminds me of the jest `exceptArrayContaining`/`ObjectContaining` pattern. This is really elegant when you read it, but turns into a gigantic mess when the expectations have changed and you're trying to find what values you forgot to update in the test cases.

As usual, if you feel more at ease with this construct, I'll update the code to meet your expectations.